### PR TITLE
Httpxe tests according to specifications.

### PR DESCRIPTION
**Pull request by a-zuckut**

### CODE REVIEW

File: specification/httpxe/src/main/resources/META-INF/services/org.kaazing.k3po.lang.el.spi.FunctionMapperSpi [hunk 1]
```diff
@@ -0,0 +1 @@
+org.kaazing.specification.httpxe.internal.Functions$Mapper
```
Comment (pkhanal) [inline specification/httpxe/src/main/resources/META-INF/services/org.kaazing.k3po.lang.el.spi.FunctionMapperSpi]: missing newline at the end of file

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.header.with.x-origin/request.rpt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+connected
+
+write method "POST"
+write version "HTTP/1.1"
+write header "HOST" "target.example.com:80"
+write header "Origin" "http://source.example.com:80"
+write header "X-Origin" "http://source.example.com:80"
+write header content-length
+write flush
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.header.with.x-origin/request.rpt]: request.with.origin.header.and.x.origin.header

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/httpxe.request.using.kct.parameter/request.rpt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+write parameter ".kct" "application/x-message-http"
+write method "POST"
+write version "HTTP/1.1"
+write header content-length
+
+write "POST HTTP/1.1\r\n"
+write "Authorization: Basic " ${httpxe:loginBase64Encoder("joe:welcome")} "\r\n"
+write "Content-Type: text/plain\r\n"
+write flush
```
Comment (a-zuckut) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/httpxe.request.using.kct.parameter/request.rpt]: Additional newline needed

File: specification/httpxe/src/test/java/org/kaazing/specification/httpxe/OriginSecurityIT.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Test
+    @Specification({
+        "request.with.kac.parameter/request",
+        "request.with.kac.parameter/response"})
+    public void shouldPassWithAccessControlWithKacParameter() throws Exception {
+        k3po.finish();
+    }
+
+}
```
Comment (a-zuckut) [inline specification/httpxe/src/test/java/org/kaazing/specification/httpxe/OriginSecurityIT.java]: Additional newline

File: specification/httpxe/src/test/java/org/kaazing/specification/httpxe/RequestsIT.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+    @Test
+    @Specification({
+        "httpxe.request.using.kct.parameter/request",
+        "httpxe.request.using.kct.parameter/response" })
+    public void shouldPassWithHttpxeReqestUsingKctParamter() throws Exception {
+        k3po.finish();
+    }
+
+}
```
Comment (a-zuckut) [inline specification/httpxe/src/test/java/org/kaazing/specification/httpxe/RequestsIT.java]: Additional newline

File: specification/httpxe/src/test/java/org/kaazing/specification/httpxe/ResponsesIT.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    @Test
+    @Specification({
+        "header.in.connection.header.but.not.body/request",
+        "header.in.connection.header.but.not.body/response"})
+    public void shouldPassWhenConnectionHeaderInHeaderNotBody() throws Exception {
+        k3po.finish();
+    }
+
+}
```
Comment (a-zuckut) [inline specification/httpxe/src/test/java/org/kaazing/specification/httpxe/ResponsesIT.java]: Additional newline

File: specification/httpxe/src/main/java/org/kaazing/specification/httpxe/internal/Functions.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        @Override
+        public String getPrefixName() {
+            return "httpxe";
+        }
+    }
+
+    private Functions() {
+        // utility
+    }
+}
```
Comment (a-zuckut) [inline specification/httpxe/src/main/java/org/kaazing/specification/httpxe/internal/Functions.java]: Additional newline

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.header/request.rpt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+connect http://localhost:8000/path
+connected
+
+write method "POST"
+write version "HTTP/1.1"
+write header "HOST" "target.example.com:80"
+write header "Origin" "http://source.example.com:80"
+write header content-length
+write flush
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.header/request.rpt]: request.with.origin.header?

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.ko.parameter/request.rpt [hunk 9]
```diff
@@ -0,0 +1,11 @@
+connect http://localhost:8000/path
+connected
+
+write parameter ".ko" "http%3A//source.example.com%3A80"
+write method "GET"
+write version "HTTP/1.1"
+write header "HOST" "target.example.com:80"
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.ko.parameter/request.rpt]: localhost:8000

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.ko.parameter/request.rpt [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+connected
+
+write parameter ".ko" "http%3A//source.example.com%3A80"
+write method "GET"
+write version "HTTP/1.1"
+write header "HOST" "target.example.com:80"
+write header "Referer" "http://target.example.com:80/any/path?query=any"
+write header content-length
+write flush
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.ko.parameter/request.rpt]: request.with.origin.query.parameter.and.referer.header?

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.ko.parameter/request.rpt [hunk 11]
```diff
@@ -0,0 +1,11 @@
+connect http://localhost:8000/path
+connected
+
+write parameter ".ko" "http%3A//source.example.com%3A80"
+write method "GET"
+write version "HTTP/1.1"
+write header "HOST" "target.example.com:80"
+write header "Referer" "http://target.example.com:80/any/path?query=any"
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.ko.parameter/request.rpt]: http://localhost:8000/any/path?query=any

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.ko.parameter/response.rpt [hunk 12]
```diff
@@ -0,0 +1,10 @@
+accept http://localhost:8000/path
+accepted
+connected
+
+read parameter ".ko" "http://source.example.com:80"
+read method "GET"
+read version "HTTP/1.1"
+read header "HOST" "target.example.com:80"
+read header "Referer" "http://target.example.com:80/any/path?query=any"
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.ko.parameter/response.rpt]: isn't this duplicate of the script above?

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.referer/request.rpt [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+connected
+
+write method "POST"
+write version "HTTP/1.1"
+write header "HOST" "target.example.com:80"
+write header "Referer" "http://target.example.com:80/different/path"
+write header "X-Origin" "http://source.examples.com:80"
+write header content-length
+write flush
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.using.referer/request.rpt]: isn't this duplicate of the script above (origin.request.using.ko.parameter)?

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.with.differing.headers/request.rpt [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+connected
+
+write method "POST"
+write version "HTTP/1.1"
+write header "HOST" "target.example.com:80"
+write header "Origin" "http://target.example.com:80"
+write header "X-Origin" "http://source.examples.com:80"
+write header content-length
+write flush
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/origin.request.with.differing.headers/request.rpt]: x.origin.header.not.identical.to.origin.header

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/request.with.kac.parameter/request.rpt [hunk 15]
```diff
@@ -0,0 +1,14 @@
+connect http://localhost:8000/path
+connected
+
+write parameter ".kac" "ex"
+write method "GET"
+write version "HTTP/1.1"
+write header "Referer" "http://source.example.com:80"
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/origin/request.with.kac.parameter/request.rpt]: do we need Referer header? 
The spec says that client can request Access-Control-* header in response by adding .kac query parameter even for same origin requests.
So, the Origin header here should be ``http://localhost:8000``

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/httpxe.request.using.http/request.rpt [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+write method "POST"
+write version "HTTP/1.1"
+write header "Content-Type" "application/x-message-http"
+write header content-length
+
+write "POST HTTP/1.1\r\n"
+write "Authorization: Basic " ${httpxe:loginBase64Encoder("joe:welcome")} "\r\n"
+write "Content-Type: text/plain"
+write flush
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/httpxe.request.using.http/request.rpt]: client.sends.wrapped.http.request?

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/httpxe.request.using.http/response.rpt [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+connected
+
+read method "POST"
+read version "HTTP/1.1"
+read header "Content-Type" "application/x-message-http"
+
+read "POST HTTP/1.1\r\n"
+read "Authorization: Basic " ${httpxe:loginBase64Encoder("joe:welcome")} "\r\n"
+read "Content-Type: text/plain"
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/httpxe.request.using.http/response.rpt]: client.sends.httpxe.request

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/httpxe.request.using.kct.parameter/request.rpt [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/httpxe.request.using.kct.parameter/request.rpt]: client.sends.httpxe.request.using.kct.parameter

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/post.request.with.km.parameter.custom/request.rpt [hunk 19]
```diff
@@ -0,0 +1,9 @@
+connect http://localhost:8000/path
+connected
+
+write parameter ".km" "(CUSTOM)"
+write method "POST"
+write version "HTTP/1.1"
+write header content-length
+write flush
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/requests/post.request.with.km.parameter.custom/request.rpt]: client.specify.custom.method.using.km.query.parameter

File: specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/responses/header.in.connection.header.but.not.body/request.rpt [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+write version "HTTP/1.1"
+write flush
+
+read status "200" "OK"
+read header "Content-Type" "application/x-message-http"
+read header "Connection" "keep-alive"
+
+read "101 Switching Protocols HTTP/1.1\r\n"
+read "Content-Type: text/html\r\n"
+
```
Comment (pkhanal) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/responses/header.in.connection.header.but.not.body/request.rpt]: I am not quite sure what is the purpose of this test case?

Comment (a-zuckut) [inline specification/httpxe/src/main/scripts/org/kaazing/specification/httpxe/responses/header.in.connection.header.but.not.body/request.rpt]: This is showing that some headers are not moved to the response body. In this case there is a Connection header in the 200 response but it does not move to the response body. (Where the 101 response is)

**Pull request merged**